### DISCUSSION TITLE
Gum Spin option to print output while executing command ?

### DISCUSSION THREAD

**Discussion by aravind-kumar** [charmbracelet/gum#149] [Q&A]
Hi 

Started using gum for my shell scripts, it's pretty cool and easy to use. 😄 

I was wondering if theres a way to print the output while the command is run as opposed to after completion.

The command takes a few minutes to complete and the user can only see the spinner .

I would like to also show the output of the command while spinner is spinning , to show progress. 
 
Today when i use --show-output option i see can see the whole output after completion of the command.
I looked at the code : [link](https://github.com/charmbracelet/gum/blob/adc82b2fce533beb3b2e787b85ab1699741a410e/spin/command.go#L34)

we seem to print the output after executing the command.

I was wondering if theres a way i could have the spinner be fixed running at the top (Like a header) while the output of command is filled below as the command is run.

I tried to play with the input command : 
- I tried to redirect the stdout and err as part of input command with "2>&1"
- I tried to rediect the output to terminal with  | tee /dev/stdout 
- The closest i could get to achieving the behaviour was passing a sleep command to spin and running the actual command in parallel 
  - With this approach i could get the output while the spinner was running for pre-determined sleep time
  - But this had its own set of issues like when the actual command failed and the spinner kept running
  - I couldnt control the spin time and the spinner kept spinning at the bottom of the screen while the output of command filled top

**Comment by wellkilo** [comment]
This is supported in current Gum. Upgrade to a recent release and use:

    gum spin --show-output --title "Working..." -- your-command arg1 arg2

You can also select streams separately:

    gum spin --show-stdout --show-stderr -- your-command

In current v2, the command writes into PTYs/buffers and the spinner's View renders those buffers while the process is running; after success the captured output is also emitted to stdout so piping still works. See [spin.go](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/spin/spin.go#L50-L96) and [the live View](https://github.com/charmbracelet/gum/blob/7179388031ae67d7f538d001be87d931f1cf5e28/spin/spin.go#L123-L145).

Your observation was correct for the older implementation linked in the question, which only printed the captured buffer after completion. The behavior has since changed.

Two caveats:

- The child may buffer its own stdout when it detects a nonstandard terminal. Use that program's unbuffered/line-buffered option if output still arrives in chunks (for Python, python -u or PYTHONUNBUFFERED=1).
- Do not add | tee as plain arguments after --; gum launches the command directly. If you need a pipeline, invoke a shell explicitly, for example sh -c 'your-command 2>&1 | tee log.txt'.
